mesonbuild/scripts/python_info.py

```diff
@@ -65,6 +65,14 @@ def links_against_libpython():
 else:
     suffix = variables.get('EXT_SUFFIX')
 
+limited_api_suffix = None
+if sys.version_info >= (3, 2):
+    try:
+        from importlib.machinery import EXTENSION_SUFFIXES
+        limited_api_suffix = EXTENSION_SUFFIXES[1]
```

Comment:
I don't believe the order in which extensions appear in this list is guaranteed by any implementation of Python. In all the versions of CPython that I could test locally, the limited API suffix always appeared second. pypy supports the limited API but does not care about the suffix and so there is only one element in this list.

There isn't, to my knowledge, a documented way to reliably get this information from different Python implementations. My first implementation used `imp.extension_suffixes()` but I moved to using `importlib.machinery` as `imp` is scheduled to be removed in 3.12. 

I would welcome any information from Python contributors about this. The current implementation seems to work for everything I have tested so far but I agree with you that this is not robust. I personally don't like this code and am eager to change it to something more robust.

Reply:
Python will happily try to load any file that happens to have an extension contained in `importlib.machinery.EXTENSION_SUFFIXES`, therefore the fact that things work if you pick any value from that list is not indicative that things are working as they should.

The only thing I find in the documentation is:
> On some platforms, Python will look for and load shared library files named with the `abi3` tag (e.g. `mymodule.abi3.so`). It does not check if such extensions conform to a Stable ABI.
which is as good as useless.

I think the only way to know what is the correct thing to do is to check what `distutils` / `setuptools` do and do the same.

@FFY00 Do you have any insight? `meson-python` checks the extension filename suffix to determine if the extension is compiled for the stable ABI also on Windows, but, accordingly to the documentation that is not guaranteed to work.

Reply:
Here is what `setuptools` does: https://github.com/pypa/setuptools/blob/80e60c3bdd5aa57f32560e5b0d946d0c07c71ac2/setuptools/command/build_ext.py#L71-L77

Reply:
Is mirroring what `setuptools` does our best option here to something as robust as possible? I do find it troubling that there does not seem to be an ironclad way to get this information.

Reply:
@amcnulty-fermat You should bring this up with the CPython developers. AFAIK there isn't any better way.

Reply:
There is [a new API design](https://github.com/python/cpython/issues/103480) for sysconfig, I commented on the [compile-specific issue](https://github.com/python/cpython/issues/103482#issuecomment-1663793710)

Reply:
> I don't believe the order in which extensions appear in this list is guaranteed by any implementation of Python. In all the versions of CPython that I could test locally, the limited API suffix always appeared second.

Yes. This is an implementation detail. On current CPython versions, and most other implementations, the suffix list is sorted by extension module lookup order.

My recommendation here would be to look for the suffix matching the following regex `^.*\.abi(?P<abi_version>[0-9]+)\.[^\.]*$`. Since the stable ABI tag is standardized on POSIX (via [PEP 3149](https://peps.python.org/pep-3149/)), this should be pretty reliable there. On Windows, even though it's not formalized anywhere, the same format is used.

That said, I don't think using `EXTENSION_SUFFIXES[1]` should be that problematic, if you wish to proceed with that instead.

In the future, this information should be provided by sysconfig.

Reply:
> On Windows, even though it's not formalized anywhere, the same format is used.

Can you please clarify this? On Windows, for Python versions 3.7 to 3.11, `importlib.machinery.EXTENSION_SUFFIXES` does not contain any suffix that contain `abi3` or similar.

Reply:
You're right, I thought that had changed at some point, but it's not the case. What changed was support in wheels, tooling should be able to handle `abi3` Windows wheels, but the extension still does not contain the `abi3` suffix indeed. Sorry for the confusion.